File: test_old/local_edge_functions/test/generate-ai-nutrition-plan.test.js
// Local test for generate-ai-nutrition-plan edge function
import { test, describe } from 'node:test';
import assert from 'node:assert';
import { generateAINutritionPlan, handleRequest } from '../functions/generate-ai-nutrition-plan.js';

// Test data - mirrors Flutter test data structure
const testMacroTargets = {
  pre_run: {
    carbs_g: 50.0,
    protein_g: 10.0,
    fat_g: 5.0,
    sodium_mg: 100.0,
    water_ml: 200.0
  },
  during_run: {
    carbs_total_g: 60.0,
    sodium_total_mg: 300.0,
    water_total_ml: 500.0
  },
  post_run: {
    carbs_g: 40.0,
    protein_g: 20.0,
    fat_g: 5.0,
    sodium_mg: 150.0,
    water_ml: 300.0
  }
};

const testNutritionPlanRequest = {
  device_id: 'test-device-local',
  age: 30,
  gender: 'female',
  weight_kg: 60.0,
  height_cm: 165.0,
  gut_training_level: 'intermediate',
  distance_miles: 6.2, // 10K
  pace_minutes_per_mile: 8.5,
  time_before_run_hours: 2.0,
  macro_targets: testMacroTargets,
  liked_foods: ['oatmeal', 'banana'],
  willing_to_try_foods: ['sports_drink'],
  disliked_foods: []
};

describe('Generate AI Nutrition Plan - Local Tests', () => {

  test('successfully generates plan with valid request', async () => {
    const result = await generateAINutritionPlan(testNutritionPlanRequest);
    
    assert.strictEqual(result.success, true, 'Should return success: true');
    assert.ok(result.plan, 'Should have plan object');
    assert.ok(result.macro_targets, 'Should have macro_targets object');
    
    // Check plan structure
    assert.ok(result.plan.before, 'Should have before phase');
    assert.ok(result.plan.during, 'Should have during phase');
    assert.ok(result.plan.after, 'Should have after phase');
    
    // Check that phases contain arrays
    assert.ok(Array.isArray(result.plan.before), 'Before phase should be array');
    assert.ok(Array.isArray(result.plan.during), 'During phase should be array');
    assert.ok(Array.isArray(result.plan.after), 'After phase should be array');
  });

  test('validates required fields', async () => {
    const invalidRequest = {
      device_id: 'test-device',
      // Missing macro_targets and other required fields
    };
    
    const result = await generateAINutritionPlan(invalidRequest);
    
    assert.strictEqual(result.success, false, 'Should return success: false');
    assert.ok(result.error, 'Should have error message');
    assert.ok(result.error.includes('Missing required fields'), 'Should mention missing fields');
  });

  test('validates macro_targets structure', async () => {
    const invalidRequest = {
      ...testNutritionPlanRequest,
      macro_targets: {
        pre_run: { carbs_g: 50 },
        // Missing during_run and post_run
      }
    };
    
    const result = await generateAINutritionPlan(invalidRequest);
    
    assert.strictEqual(result.success, false, 'Should return success: false');
    assert.ok(result.error, 'Should have error message');
    assert.ok(result.error.includes('Missing macro_targets'), 'Should mention missing macro_targets phases');
  });

  test('respects food preferences', async () => {
    const requestWithPrefs = {
      ...testNutritionPlanRequest,
      liked_foods: ['oatmeal', 'banana'],
      disliked_foods: ['sports_drink']
    };
    
    const result = await generateAINutritionPlan(requestWithPrefs);
    
    assert.strictEqual(result.success, true, 'Should return success: true');
    
    // Check that disliked foods are not included
    const allFoods = [
      ...result.plan.before,
      ...result.plan.during,
      ...result.plan.after
    ];
    
    const hasSportsDrink = allFoods.some(food => 
      food.food_name && food.food_name.toLowerCase().includes('sports drink')
    );
    
    assert.strictEqual(hasSportsDrink, false, 'Should not include disliked sports drink');
  });

  test('generates foods with proper serving sizes', async () => {
    const result = await generateAINutritionPlan(testNutritionPlanRequest);
    
    assert.strictEqual(result.success, true, 'Should return success: true');
    
    // Check all food items for proper serving increments
    const allFoods = [
      ...result.plan.before,
      ...result.plan.during,
      ...result.plan.after
    ];
    
    for (const food of allFoods) {
      assert.ok(food.servings, 'Should have servings');
      assert.ok(food.food_name, 'Should have food_name');
      assert.ok(food.description, 'Should have description');
      
      // Check servings are in 0.5 increments
      const remainder = (food.servings * 2) % 1;
      assert.ok(Math.abs(remainder) < 0.01, 
        `Food "${food.food_name}" has servings of ${food.servings} - should be in 0.5 increments`);
      
      // Check minimum serving
      assert.ok(food.servings >= 0.5, 
        `Food "${food.food_name}" has servings of ${food.servings} - should be at least 0.5`);
    }
  });

  test('HTTP handler works correctly', async () => {
    const response = await handleRequest(testNutritionPlanRequest);
    
    assert.strictEqual(response.status, 200, 'Should return 200 status');
    assert.ok(response.data, 'Should have data object');
    assert.strictEqual(response.data.success, true, 'Data should indicate success');
  });

  test('HTTP handler handles errors gracefully', async () => {
    const invalidRequest = { device_id: 'test' }; // Missing required fields
    
    const response = await handleRequest(invalidRequest);
    
    assert.ok(response.status >= 400, 'Should return error status');
    assert.ok(response.data, 'Should have data object');
    assert.strictEqual(response.data.success, false, 'Data should indicate failure');
    assert.ok(response.data.error, 'Should have error message');
  });

  test('handles empty food preferences', async () => {
    const requestWithEmptyPrefs = {
      ...testNutritionPlanRequest,
      liked_foods: [],
      willing_to_try_foods: [],
      disliked_foods: []
    };
    
    const result = await generateAINutritionPlan(requestWithEmptyPrefs);
    
    assert.strictEqual(result.success, true, 'Should handle empty preferences');
    assert.ok(result.plan, 'Should still generate a plan');
  });

  test('phase constraints are respected', async () => {
    const result = await generateAINutritionPlan(testNutritionPlanRequest);
    
    assert.strictEqual(result.success, true, 'Should return success: true');
    
    // Check that during phase only has during-appropriate foods
    const duringFoods = result.plan.during;
    
    for (const food of duringFoods) {
      const foodName = food.food_name.toLowerCase();
      
      // Should not have pre-run only foods during run
      assert.ok(!foodName.includes('oatmeal'), 
        'Oatmeal should not appear during runs - safety constraint');
    }
    
    // Should have some suitable during-run foods if available
    const hasDuringFoods = duringFoods.some(food => {
      const foodName = food.food_name.toLowerCase();
      return foodName.includes('sports') || foodName.includes('gel') || foodName.includes('energy');
    });
    
    // Only check if we have during foods (depends on targets)
    if (duringFoods.length > 0) {
      assert.ok(hasDuringFoods, 'Should include suitable during-run foods when needed');
    }
  });

});

// Manual test runner for development
if (import.meta.url === `file://${process.argv[1]}`) {
  console.log('Running manual test...');
  
  const result = await generateAINutritionPlan(testNutritionPlanRequest);
  console.log('Result:', JSON.stringify(result, null, 2));
  
  console.log('\nTest completed successfully!');
}

File: test_old/integration/edge_functions/generate_ai_nutrition_plan_test.dart
/// Generate AI Nutrition Plan Edge Function Tests
/// 
/// These tests validate the AI nutrition plan generation edge function
/// and its ability to meet macro targets within tight tolerances.
library;

import 'package:flutter_test/flutter_test.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

import '../../helpers/test_data.dart';
import '../../helpers/test_config.dart';

void main() {
  group('Generate AI Nutrition Plan Function', () {
    late SupabaseClient supabase;

    setUpAll(() {
      // Validate test environment before running
      final configError = TestConfig.validateConfiguration();
      if (configError != null) {
        fail('Test configuration invalid: $configError');
      }

      supabase = SupabaseClient(
        TestConfig.supabaseUrl,
        TestConfig.supabaseAnonKey,
      );
    });

    testWidgets('successfully generates plan with adjusted macro targets', 
        (tester) async {
      // Skip if network tests disabled
      if (!TestConfig.enableNetworkTests) {
        return;
      }

      // Arrange: Create request with user-adjusted macro targets
      final requestData = TestData.testNutritionPlanRequest;

      // Act: Call the edge function
      final response = await supabase.functions.invoke(
        'generate-ai-nutrition-plan',
        body: requestData,
      ).timeout(TestConfig.networkTestTimeout);

      // Assert: Verify successful response
      expect(response.status, lessThan(400), 
          reason: 'Edge function should succeed. Status: ${response.status}, Response: ${response.data}');

      final responseData = response.data as Map<String, dynamic>;
      
      // Verify response structure
      expect(responseData, containsPair('success', true));
      expect(responseData, contains('plan'));
      expect(responseData, contains('macro_targets'));
      expect(responseData, contains('detailed_message'));

      // Extract and validate plan structure
      final plan = responseData['plan'] as Map<String, dynamic>;
      expect(plan, contains('before'));
      expect(plan, contains('during'));
      expect(plan, contains('after'));

      // Validate each phase has food items
      final beforeItems = plan['before'] as List<dynamic>;
      final duringItems = plan['during'] as List<dynamic>;
      final afterItems = plan['after'] as List<dynamic>;

      expect(beforeItems, isNotEmpty, reason: 'Should have before-run food items');
      expect(duringItems, isNotEmpty, reason: 'Should have during-run food items for 10K');
      expect(afterItems, isNotEmpty, reason: 'Should have after-run food items');

      // Validate food item structure
      for (final item in beforeItems) {
        final itemMap = item as Map<String, dynamic>;
        expect(itemMap, contains('food_name'));
        expect(itemMap, contains('servings'));
        expect(itemMap, contains('carbs_per_serving'));
        expect(itemMap['servings'], isA<num>());
        expect(itemMap['carbs_per_serving'], isA<num>());
      }

      // CRITICAL TEST: Verify adjusted macro targets are preserved
      final returnedMacros = responseData['macro_targets'] as Map<String, dynamic>;
      final returnedPreRun = returnedMacros['pre_run'] as Map<String, dynamic>;
      final returnedDuringRun = returnedMacros['during_run'] as Map<String, dynamic>;
      final returnedPostRun = returnedMacros['post_run'] as Map<String, dynamic>;

      final originalMacros = TestData.testMacroTargets;
      
      expect(returnedPreRun['carbs_g'], equals(originalMacros['pre_run']!['carbs_g']),
          reason: 'Pre-run carbs should match adjusted value');
      expect(returnedPreRun['protein_g'], equals(originalMacros['pre_run']!['protein_g']),
          reason: 'Pre-run protein should match adjusted value');
      expect(returnedDuringRun['carbs_total_g'], equals(originalMacros['during_run']!['carbs_total_g']),
          reason: 'During-run carbs should match adjusted value');
      expect(returnedPostRun['carbs_g'], equals(originalMacros['post_run']!['carbs_g']),
          reason: 'Post-run carbs should match adjusted value');
    });

    testWidgets('validates food recommendations align with macro targets', 
        (tester) async {
      if (!TestConfig.enableNetworkTests) return;

      // Arrange: Request with specific macro expectations
      final requestData = TestData.testNutritionPlanRequest;
      final expectedMacros = TestData.testMacroTargets;

      // Act: Generate plan
      final response = await supabase.functions.invoke(
        'generate-ai-nutrition-plan',
        body: requestData,
      ).timeout(TestConfig.networkTestTimeout);

      expect(response.status, lessThan(400));
      final responseData = response.data as Map<String, dynamic>;
      final plan = responseData['plan'] as Map<String, dynamic>;

      // Assert: Calculate actual macros from food recommendations
      final phases = ['before', 'during', 'after'];
      final phaseKeys = ['pre_run', 'during_run', 'post_run'];
      
      for (int i = 0; i < phases.length; i++) {
        final phaseName = phases[i];
        final phaseKey = phaseKeys[i];
        final items = plan[phaseName] as List<dynamic>;
        
        double totalCarbs = 0.0;
        double totalProtein = 0.0;
        
        for (final item in items) {
          final itemMap = item as Map<String, dynamic>;
          final servings = (itemMap['servings'] as num).toDouble();
          final carbsPerServing = (itemMap['carbs_per_serving'] as num? ?? 0).toDouble();
          final proteinPerServing = (itemMap['protein_per_serving'] as num? ?? 0).toDouble();
          
          totalCarbs += servings * carbsPerServing;
          totalProtein += servings * proteinPerServing;
        }

        // Validate carb targets (within tolerance)
        final expectedCarbs = phaseKey == 'during_run' 
            ? expectedMacros[phaseKey]!['carbs_total_g']!
            : expectedMacros[phaseKey]!['carbs_g']!;
            
        expect(totalCarbs, closeTo(expectedCarbs, TestConfig.carbToleranceGrams),
            reason: '$phaseName carbs (${totalCarbs.toStringAsFixed(1)}g) should be close to target ($expectedCarbs g)');

        // Validate protein targets for phases that have protein targets
        if (phaseKey != 'during_run') {
          final expectedProtein = expectedMacros[phaseKey]!['protein_g']!;
          expect(totalProtein, closeTo(expectedProtein, TestConfig.proteinToleranceGrams),
              reason: '$phaseName protein (${totalProtein.toStringAsFixed(1)}g) should be close to target ($expectedProtein g)');
        }
      }
    });

    testWidgets('meets tightened nutrient targets within tolerance', 
        (tester) async {
      if (!TestConfig.enableNetworkTests) return;

      // Arrange: Request with specific macro targets
      final requestData = TestData.testNutritionPlanRequest;
      final expectedMacros = TestData.testMacroTargets;

      // Act: Generate plan
      final response = await supabase.functions.invoke(
        'generate-ai-nutrition-plan',
        body: requestData,
      ).timeout(TestConfig.networkTestTimeout);

      expect(response.status, lessThan(400));
      final responseData = response.data as Map<String, dynamic>;
      final plan = responseData['plan'] as Map<String, dynamic>;

      // Assert: Verify improved tolerances for carbs, sodium, and fluids
      final phases = [
        {'name': 'before', 'key': 'pre_run'},
        {'name': 'during', 'key': 'during_run'},
        {'name': 'after', 'key': 'post_run'}
      ];
      
      for (final phase in phases) {
        final phaseName = phase['name'] as String;
        final phaseKey = phase['key'] as String;
        final items = plan[phaseName] as List<dynamic>;
        
        double totalCarbs = 0.0;
        double totalSodium = 0.0;
        double totalFluids = 0.0;
        
        for (final item in items) {
          final itemMap = item as Map<String, dynamic>;
          final servings = (itemMap['servings'] as num).toDouble();
          
          // Handle different property names across phases
          final carbsKey = phaseKey == 'during_run' ? 'carbs_grams' : 'carbs_grams';
          final sodiumKey = 'sodium_mg';
          final fluidsKey = 'fluids_ml';
          
          totalCarbs += (itemMap[carbsKey] as num? ?? 0).toDouble();
          totalSodium += (itemMap[sodiumKey] as num? ?? 0).toDouble();
          totalFluids += (itemMap[fluidsKey] as num? ?? 0).toDouble();
        }

        // Get expected targets
        final expectedCarbs = phaseKey == 'during_run' 
            ? expectedMacros[phaseKey]!['carbs_total_g']!
            : expectedMacros[phaseKey]!['carbs_g']!;
        final expectedSodium = phaseKey == 'during_run'
            ? expectedMacros[phaseKey]!['sodium_total_mg']!
            : expectedMacros[phaseKey]!['sodium_mg']!;
        final expectedFluids = phaseKey == 'during_run'
            ? expectedMacros[phaseKey]!['water_total_ml']!
            : expectedMacros[phaseKey]!['water_ml']!;

        // CRITICAL: Verify tightened tolerances
        expect(totalCarbs, closeTo(expectedCarbs, 10.0),
            reason: '$phaseName carbs (${totalCarbs.toStringAsFixed(1)}g) should be within ±10g of target (${expectedCarbs}g)');
            
        expect(totalSodium, closeTo(expectedSodium, 20.0),
            reason: '$phaseName sodium (${totalSodium.toStringAsFixed(0)}mg) should be within ±20mg of target (${expectedSodium}mg)');
            
        expect(totalFluids, closeTo(expectedFluids, 20.0),
            reason: '$phaseName fluids (${totalFluids.toStringAsFixed(0)}ml) should be within ±20ml of target (${expectedFluids}ml)');
      }
    });

    testWidgets('generates clean serving sizes in 0.5 increments', 
        (tester) async {
      if (!TestConfig.enableNetworkTests) return;

      // Arrange: Standard request
      final requestData = TestData.testNutritionPlanRequest;

      // Act: Generate plan
      final response = await supabase.functions.invoke(
        'generate-ai-nutrition-plan',
        body: requestData,
      ).timeout(TestConfig.networkTestTimeout);

      expect(response.status, lessThan(400));
      final responseData = response.data as Map<String, dynamic>;
      final plan = responseData['plan'] as Map<String, dynamic>;

      // Assert: All servings should be in 0.5 increments
      final allItems = [
        ...(plan['before'] as List<dynamic>),
        ...(plan['during'] as List<dynamic>),
        ...(plan['after'] as List<dynamic>),
      ];

      for (final item in allItems) {
        final itemMap = item as Map<String, dynamic>;
        final servings = (itemMap['servings'] as num).toDouble();
        final foodName = itemMap['food_name'] as String;
        
        // Check that servings are in 0.5 increments
        final remainder = (servings * 2) % 1;
        expect(remainder, closeTo(0.0, 0.01),
            reason: 'Food "$foodName" has servings of $servings - should be in 0.5 increments (0.5, 1.0, 1.5, 2.0, etc.)');
            
        // Ensure minimum serving of 0.5
        expect(servings, greaterThanOrEqualTo(0.5),
            reason: 'Food "$foodName" has servings of $servings - should be at least 0.5');
            
        // Verify description contains clean amounts (no decimals like 1.73)
        final description = itemMap['description'] as String;
        final match = RegExp(r'^(\d+(?:\.\d+)?)\s').firstMatch(description);
        if (match != null) {
          final amountStr = match.group(1)!;
          final amount = double.parse(amountStr);
          final amountRemainder = (amount * 2) % 1;
          expect(amountRemainder, closeTo(0.0, 0.01),
              reason: 'Food "$foodName" description "$description" contains non-standard amount $amount');
        }
      }
    });

    testWidgets('consolidates duplicate foods in plan', 
        (tester) async {
      if (!TestConfig.enableNetworkTests) return;

      // Arrange: Request that might generate duplicate foods
      final requestData = TestData.testNutritionPlanRequest.cast<String, dynamic>();
      // Use preferences that might cause same food to be selected multiple times
      requestData['liked_foods'] = ['sports_drink', 'energy_gel', 'banana'];

      // Act: Generate plan
      final response = await supabase.functions.invoke(
        'generate-ai-nutrition-plan',
        body: requestData,
      ).timeout(TestConfig.networkTestTimeout);

      expect(response.status, lessThan(400));
      final responseData = response.data as Map<String, dynamic>;
      final plan = responseData['plan'] as Map<String, dynamic>;

      // Assert: No duplicate foods within each phase
      final phases = ['before', 'during', 'after'];
      
      for (final phaseName in phases) {
        final items = plan[phaseName] as List<dynamic>;
        final foodNames = <String>[];
        
        for (final item in items) {
          final itemMap = item as Map<String, dynamic>;
          final foodName = itemMap['food_name'] as String;
          
          // Remove any boost labels for comparison
          final baseName = foodName.replaceAll(RegExp(r' \((sodium|carb|hydration) boost\)$'), '');
          
          expect(foodNames, isNot(contains(baseName)),
              reason: 'Duplicate food found in $phaseName phase: $baseName. Foods: ${foodNames.join(', ')}');
              
          foodNames.add(baseName);
        }
      }
    });

    testWidgets('respects food preferences in AI-generated plan', 
        (tester) async {
      if (!TestConfig.enableNetworkTests) return;

      // Arrange: Request with specific food preferences
      final requestData = TestData.testNutritionPlanRequest.cast<String, dynamic>();
      requestData['liked_foods'] = ['oatmeal', 'gels', 'protein_shake'];
      requestData['disliked_foods'] = ['coffee', 'banana', 'coconut_water'];

      // Act: Generate plan
      final response = await supabase.functions.invoke(
        'generate-ai-nutrition-plan',
        body: requestData,
      ).timeout(TestConfig.networkTestTimeout);

      expect(response.status, lessThan(400));
      final responseData = response.data as Map<String, dynamic>;
      final plan = responseData['plan'] as Map<String, dynamic>;

      // Assert: Check food preferences are respected
      final allItems = [
        ...(plan['before'] as List<dynamic>),
        ...(plan['during'] as List<dynamic>),
        ...(plan['after'] as List<dynamic>),
      ];

      final allFoodNames = allItems
          .map((item) => (item as Map<String, dynamic>)['food_name'] as String)
          .toSet();

      // Should not include disliked foods
      final dislikedFoods = requestData['disliked_foods'] as List<String>;
      for (final dislikedFood in dislikedFoods) {
        expect(allFoodNames, isNot(contains(dislikedFood)),
            reason: 'Should not include disliked food: $dislikedFood');
      }

      // Should prefer liked foods when phase-appropriate
      final likedFoods = requestData['liked_foods'] as List<String>;
      int likedFoodsUsed = 0;
      for (final likedFood in likedFoods) {
        if (allFoodNames.contains(likedFood)) {
          likedFoodsUsed++;
        }
      }

      expect(likedFoodsUsed, greaterThan(0),
          reason: 'Should include at least some liked foods when appropriate');
    });

    testWidgets('enforces food phase safety constraints', 
        (tester) async {
      if (!TestConfig.enableNetworkTests) return;

      // Arrange: Request with foods that have phase constraints
      final requestData = TestData.testNutritionPlanRequest.cast<String, dynamic>();
      requestData['liked_foods'] = ['oatmeal', 'banana', 'gels', 'sports_drink'];
      requestData['distance_miles'] = 10.0; // Long enough to require during-run fueling

      // Act: Generate plan
      final response = await supabase.functions.invoke(
        'generate-ai-nutrition-plan',
        body: requestData,
      ).timeout(TestConfig.networkTestTimeout);

      expect(response.status, lessThan(400));
      final responseData = response.data as Map<String, dynamic>;
      final plan = responseData['plan'] as Map<String, dynamic>;

      // Assert: Verify phase constraints are enforced
      final duringItems = plan['during'] as List<dynamic>;
      final duringFoodNames = duringItems
          .map((item) => (item as Map<String, dynamic>)['food_name'] as String)
          .toSet();

      // CRITICAL: Pre-run only foods should never appear during runs
      expect(duringFoodNames, isNot(contains('oatmeal')),
          reason: 'Oatmeal should never appear during runs - safety constraint');
      expect(duringFoodNames, isNot(contains('banana')),
          reason: 'Banana should never appear during runs - safety constraint');

      // During-run suitable foods should be preferred
      final suitableDuringFoods = duringFoodNames
          .where((food) => ['gels', 'sports_drink', 'energy_chews', 'electrolyte_tablets']
              .any((suitable) => food.toLowerCase().contains(suitable.toLowerCase())))
          .toList();
          
      expect(suitableDuringFoods, isNotEmpty,
          reason: 'Should include suitable during-run foods');
    });
  });
}